File: Arrays_and_Hashing/C++/36_valid_sudoku.cpp
/*
 * RUNTIME: 22 ms - Beats 38.46% of submissions
 * MEMORY - 19.53 MB - Beats 53.09% of submissions
 */

#include <algorithm>
#include <map>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

/* Thoughtprocess:
 *  1. Create enum for better readability and avoid three different maps
 *  2. Check vor Validy of Sudoku for rows and columns
 *  3. Check Validy of Sudoku for subboxes
 *  4. If Program reaches end of function, Sudoku is valid
 */

class Solution
{
public:

    enum sudokuConditions // Avoid Magic Numbers
    {
        row = 0,
        column = 1,
        subBox = 2
    };

    bool isValidSudoku(vector<vector<char>> &board)
    {
        unordered_map<char, int> validyCondition[3]; 

        for (int i = 0; i < 9; i++)
        {
            validyCondition[row].clear();
            validyCondition[column].clear();
            for (int j = 0; j < 9; j++)
            {
                if (board[i][j] != '.')
                {
                    validyCondition[row][board[i][j]]++;
                }
                if (board[j][i] != '.')
                {
                    validyCondition[column][board[j][i]]++;
                }

                //Check Condition every iteration
                if (validyCondition[row][board[i][j]] > 1)
                    return false;
                if (validyCondition[column][board[j][i]] > 1)
                    return false;
            }
        }

        //Create for Loop to define Box
        for (int i = 0; i < 9; i += 3)
            for (int j = 0; j < 9; j += 3)
            {
                validyCondition[subBox].clear();
                //Create for Loop to check SubBox
                for (int x = i; x < i + 3; x++)
                    for (int y = j; y < j + 3; y++)
                    {
                        if (board[x][y] != '.')
                            validyCondition[subBox][board[x][y]]++;
                        if (validyCondition[subBox][board[x][y]] > 1)
                            return false;
                    }
            }
        return true;
    }
};

File: Arrays_and_Hashing/Python/36_valid_sudoku.py
# RUNTIME: 116 ms, faster than 15.14% of python3 Submissions.
# MEMORY USAGE: 16.55 MB, less than 52.29% of python3 Submissions.

from __future__ import annotations
from ast import *
from collections import Counter


# My thougtprocess:
# 1. Check store numeric values in a list for Columns and Rows
# 2. Check store numeric values in a list for sub-boxes
# 3. Check if the list is empty then it is a valid sudoku
# 4. Check if the most common element in the list is greater than 1 then it is not a valid sudoku
# 5. Else it is a valid sudoku

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        s = []
        c = ""
        c2 = ""

        # check rows and columns
        for i in range(9):
            c = ""
            for j in range(9):
                #check rows
                if board[i][j] == ".":
                    pass
                else:
                    c += str(board[i][j])
                #check columns
                if board[j][i]  == ".":
                    pass
                else:
                    c2 += str(board[j][i])
                    
            s.append(list(c))
            s.append(list(c2))

        # check sub-boxes
        for i in range(0, 9, 3):
            for j in range(0, 9, 3):
                c = ""
                for k in range(3):
                    for l in range(3):
                        if board[i + k][j + l] == ".":
                            pass
                        else:
                            c += str(board[i + k][j + l])
                s.append(list(c))

        # if list is empty then it is a valid sudoku
        if s.count([]) == 27:
            return True

        # if most common is greater than 1 then it is not a valid sudoku
        for x in s:
            if x == []:
                continue
            cnt = Counter(x)
            if cnt.most_common(1)[0][1] > 1:
                return False

        return True
